### PR TITLE
Region Picker concurrency overhaul

### LINKED ISSUE #633: Filter out inactive regions from the app's regions list
We shouldn't show inactive regions like Rogue Valley

### PR DESCRIPTION
**Pull request by ualch9**
## OBAKit Changes
- New RegionPicker onboarding view (and, replaces RegionPickerViewController).
  - Add a "preview" map, for additional context on what a region is. If the view gets too distracting, it may be collapsed.
  - Add a inactive/experimental regions filter (fixes #633).
- New CustomRegionForm (replaces CustomRegionBuilderController, etc.).
  - Add `RenamableNavigationTitle`, a polyfill for iOS 16's `navigationTitle` editing.
- Add `View.errorAlert` for displaying an error alert, regardless of whether the error is localized or not.
- Add `LocalizedAlertError`, for creating a LocalizedAlert for `is LocalizedAlert` or `is not LocalizedAlert`.

### UI recordings
| Map | Filter | Custom Region |
| --- | --- | --- |
| <video src="https://user-images.githubusercontent.com/22162410/215191753-59da42cf-f229-414e-88d0-843569c89b9b.mov"> | <video src="https://user-images.githubusercontent.com/22162410/215191900-2c9b1cc0-644a-4f35-b2a9-918f7e171599.mov"> | <video src="https://user-images.githubusercontent.com/22162410/215192582-d951b3e2-42dc-4a3f-a608-50b48c066607.mov"> |

A delete action appears when *editing* a region. The delete button is disabled if it is the currently selected region.
Screenshot 2023-01-27 at 9 40 43 PM

`Automatically choose region` makes the region selection and disables the Picker.
Screenshot 2023-01-27 at 10 40 40 PM

### About `LocalizedAlertError`
Normally, a `is not LocalizedAlert` will have an error description of just `This operation couldn't be completed`. `LocalizedAlertError` will instead create an error description using the name of the specific Error class/struct/enum, this makes it easier for developers to pinpoint the error location.
Screenshot demostrating LocalizedAlertError creating an error description using the Error's type name

## OBAKitCore changes
- Updated `RegionAPIService` to Concurrency
- API response handling code is moved from `class RESTAPIService` to `any [protocol] APIService`.
  - Add `APIService.getData(for:)`
  - Add `APIService.getData(for:decodeAs:using:)`

## OBAKitTests changes
- Overhaul RegionAPIService unit tests (tldr: use async and swift enums, instead of `waitUntil`).
- Added `DelegateTestingHelper.DidCallDelegateMethod` to `OBAKitTests`.

## Old recording
https://user-images.githubusercontent.com/22162410/212872734-5c3fcfbd-ac81-4da4-98eb-e4e78670d1bf.mov

### CODE REVIEW

**Comment by ualch9** [comment]
Hm, want to try one more thing with one-time location authorization...

**Comment by ualch9** [comment]
> Hm, want to try one more thing with one-time location authorization...

Never mind, ready for review.

**Comment by aaronbrethorst** [comment]
@ualch9 can you take a quick pass through the commits here and squash where appropriate?

**Comment by ualch9** [comment]
Rebased, added some comments, and fixed some localization. I hope this helps with your review.

**Comment by aaronbrethorst** [comment]
@ualch9 awesome, I'll review today.

File: OBAKitCore/Network/APIService.swift [hunk 1]
```diff
@@ -6,9 +6,13 @@
 //  This source code is licensed under the Apache 2.0 license found in the
 //  LICENSE file in the root directory of this source tree.
 //
+
+import os.log
+
 public protocol APIService {
     var configuration: APIServiceConfiguration { get }
     var dataLoader: URLDataLoader { get }
+    var logger: os.Logger { get }
```
[the remark this answers is not in the document]
Comment (aaronbrethorst) [reply OBAKitCore/Network/APIService.swift:15]: 👍

File: OBAKitTests/Modeling/Regions Model Service Tests/RegionsModelOperationTests.swift [hunk 2]
```diff
@@ -17,78 +17,67 @@ import MapKit
...
-                    expect(pugetSound.serviceRect.maxY).to(beCloseTo(mapRect.maxY))
-
-                    expect(pugetSound.centerCoordinate.latitude).to(beCloseTo(47.795091214055))
-                    expect(pugetSound.centerCoordinate.longitude).to(beCloseTo(-122.49868405298474))
-
-                    done()
-                }
-            }
-        }
+        let response = try await regionsAPIService.getRegions(apiPath: regionsAPIPath)
```
Comment (aaronbrethorst) [inline OBAKitTests/Modeling/Regions Model Service Tests/RegionsModelOperationTests.swift:24]: 🎉

**Review by aaronbrethorst — CHANGES_REQUESTED** [review]
Looks great! I noticed a couple small UI changes and had a couple pieces of feedback/questions for you. Looking forward to getting this shipped. it's super cool.

File: OBAKit/Onboarding/RegionPicker/RegionCustomForm.swift [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+//  Created by Alan Chu on 1/20/23.
+//
+
+import MapKit
+import SwiftUI
+import OBAKitCore
+import CoreLocationUI
+
+/// Create (or edit) a custom region.
+struct RegionCustomForm: View {
```
[the remark this answers is not in the document]
Comment (ualch9) [reply OBAKit/Onboarding/RegionPicker/RegionCustomForm.swift:14]: #635

File: OBAKitCore/Location/Regions/RegionsService.swift [hunk 4]
```diff
@@ -338,43 +335,38 @@ public class RegionsService: NSObject, LocationServiceDelegate {
...
-
-            switch result {
-            case .failure(let error):
-                self.notifyDelegatesDisplayError(error)
-            case .success(let response):
-                guard response.list.count > 0 else { return }
-                self.regions = response.list
-                self.updateCurrentRegionFromLocation()
-            }
+        // FIXME: Audit which delegates are doing stuff on background thread, when they should be on Main.
```
[the remark this answers is not in the document]
Comment (ualch9) [reply OBAKitCore/Location/Regions/RegionsService.swift:348]: #635

File: OBAKit/Onboarding/RegionPicker/RegionCustomForm.swift [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    // MARK: Other Form state
+
+    @State private var disableForm: Bool = false
+    @State private var isPresentingDeleteConfirmation = false
+    @State private var error: Error?
+
+    var body: some View {
+        Form {
+            Section(OBALoc("custom_region_builder_controller.base_url_section.header_title", value: "Base URL", comment: "Title of the Base URL header.")) {
+                TextField("https://api.tampa.onebusaway.org/api/", text: $baseURLString)
```
[the remark this answers is not in the document]
Comment (ualch9) [reply OBAKit/Onboarding/RegionPicker/RegionCustomForm.swift:42]: agh!

**Comment by aaronbrethorst** [comment]
Awesome thanks @ualch9!

**Pull request merged**